### ISSUE TITLE
dash-mapd-demo application error

### ISSUE THREAD

**Issue by cldougl** [plotly/dash-sample-apps#333]
# App name
dash-mapd-demo
# Description of bug
https://dash-gallery.plotly.host/dash-mapd-demo/ application error
More information in application logs: https://dash-gallery.plotly.host/Manager/apps/dash-mapd-demo/logs
# How to replicate the bug
Visit https://dash-gallery.plotly.host/dash-mapd-demo/
# Other notes

**Comment by cldougl** [comment]
Suggested resolution:
1. add a cached fallback 
2. check if anything happened to the dB recently 
3. add retry logic around the connection

**Comment by tarzzz** [comment]
> check if anything happened to the dB recently

I checked some logs, and it seems every time we try to access the DB, it crashes the container leading to container restart. I suspect this could be because this DB has `7009728` records (7M+) and not enough resources to handle that level of data.

I would recommend coming up with a smaller dataset (or use postgres/sqlite if that works) for this app.

**Comment by cldougl** [comment]
> I suspect this could be because this DB has 7009728 records (7M+) and not enough resources to handle that level of data
@tarzzz why not allot more resources? (ie is this a pricing issue)
This app is supposed to demo the Omnisci DB connection so I don't think it quite resolves the problem to use postgres/sqlite

**Comment by tarzzz** [comment]
> This app is supposed to demo the Omnisci DB connection so I don't think it quite resolves the problem to use postgres/sqlite

👍 . The top-comment ("add a cached fallback") suggested using postgres/sqlite so I thought that would be ok.

> why not allot more resources? (ie is this a pricing issue)

The records are loaded to the DB on container creation, so increasing the number of containers will not help to decrease the load. Currently each running container in the cluster has 7M records stored in it, which is very inefficient.

IMHO it would be better to use a smaller dataset (easier to maintain and better performance-wise) unless the app requires to query (or use) the 7M records for its functioning..

**Comment by josegonzalez** [comment]
@tarzzz is the omnisci database still running, or is it crashing?

**Comment by bpostlethwaite** [comment]
Any chance we can get an update on the status of this one? If we don't have time to fix it properly I think @cldougl is right, let's just pay more and create an issue to cut costs later.

**Comment by tarzzz** [comment]
@bpostlethwaite The PR to use the smaller dataset is ready to be merged: https://github.com/plotly/dash-sample-apps/pull/339/

I am waiting on a 💃 on that one.

> is the omnisci database still running, or is it crashing?

The old database is running, but it crashes as soon as a query is made on it. The new database (with more RAM and smaller dataset) is available now.. @ycaokris will update her app to use the newer database and we can monitor the performance then..

**Comment by tarzzz** [comment]
The PR has been merged. I will wait for @ycaokris to verify that her app is working okay now before closing the issue.

**Comment by bpostlethwaite** [comment]
Thanks for the update @tarzzz !

**Comment by ycaokris** [comment]
@tarzzz  thanks for updating database. Confirmed that this table has already been added. 👍  I will make adaptations in the app.

I took a look at #339 and noticed that you've downgraded Omnisci Docker image to an older version(v4.5 was released 8 months ago and now it's 4.8). 
This is not compatible with python client I'm using in the app and I have to downgrade. i.e. Downgrading `pymapd` to a matched version(0.11.4 - > **<0.10.0**) for v4.5 DB. You may find changes in https://github.com/omnisci/pymapd/releases stating this version incompatibility. Although I could still run my app, being limited to use an older python client skipped some features, for example, 

**Add ability to login using sessionid** in v0.12.0,
**Ensure execute() 'operation' argument trimmed before sending** in v 0.16.0. 

These are not must-have feature in my app, but for continuing follow up and maximizing possible errors in this app, I'd like to know is there a reason we have to use this Image version, not fixed at a more recent one?

This smaller dataset, flight_10k_2008, is still existing in the docker image you installed prior to #339 in `os-cpu`, so I'm confused why we change our image. Check
https://www.omnisci.com/platform/downloads/OS-installation-recipes?compute_mode=cpuonly&environment=docker
<img width="828" alt="Screen Shot 2019-11-05 at 11 23 21 AM" src="https://user-images.githubusercontent.com/35306149/68234101-90407500-ffce-11e9-93fe-47afb85bc552.png">

**Comment by ycaokris** [comment]
Updates for this app fix Nov 8: 

#335 has done the following changes:

- Use dataset from #339
- Retry once if initial connection loses. 
- Close all db_connect() after every usage.
- Return empty chart as one fallback solution if any individual connection fails, we'll get a valid app as following output:

<img width="1307" alt="Screen Shot 2019-11-08 at 1 25 04 PM" src="https://user-images.githubusercontent.com/35306149/68503007-d21c2600-022f-11ea-95dd-b44ae9c8a40e.png">

 I've thought about local cache but it might be better to just reveal empty chart to alert a db fail, rather than always leave a cached, usable app. @cldougl  Let me know how you find about this. 

Thanks @tarzzz  for making db changes. After #335 has been merged/deployed,  I'd like another issue for performance follow-up(like resource usage, container stats) record for this new instance.

**Comment by tarzzz** [comment]
> noticed that you've downgraded Omnisci Docker image to an older version(v4.5 was released 8 months ago and now it's 4.8).

Yes. Well, technically I did not downgrade it.. : ) I just pinned it to the one which we were using when the Omnisci DB was deployed for the first time..

I tried the docker image with the `latest` tag, and found it to be unreliable. The database server would hang during startup causing the test data not being loaded. I haven't investigated the issue with their official docker image, but if we do want to use it the next steps would be to log an issue with their org and work out on the resolution.. 

@ycaokris Please let me know if this version is ok with you (we can close this issue) or if you want me to look more into the upgrading.

**Comment by ycaokris** [comment]
This version works with me and app functionality isn't getting affected so far. That is, I don't need to upgrade for current demo.  Without a detailed information/log on how this database crashes I'd stay with what we have right now.
Just found out that Omnisci has a [community forum](https://community.omnisci.com/home) from which we could further take a look if any database error occurs.

**Comment by ycaokris** [comment]
Updates for this issue:

Merged #335 for fallback adaptation to new instance. A valid app is able to run with same logic. 

![iphone](https://user-images.githubusercontent.com/35306149/68502892-8bc6c700-022f-11ea-8aba-8640cf590f11.gif)

Any future connection error will leads to empty app layout and I'll set up an instance check up with @tarzzz at least one time after two weeks. 

Push to gallery and restore on portal  https://dash-gallert.plotly.host/dash-mapd-demo/
